### PR TITLE
Add matmul optimization for the case A.ndim <= 2 && B.ndim >= 3

### PR DESCRIPTION
**Pull request by skrah**
This addresses #18862.

### CODE REVIEW

**Comment by skrah** [comment]
This PR implements more or less @colesbury's suggestion **C** from #18862.  Suggestion **C** ends up calling the existing optimization for `A.ndim >= 3 && B.ndim <= 2` after transposing the inner dimensions of the arguments and swapping them.

***Memory usage***

Memory usage as reported by `valgrind --tool=massif`.

**Test case**

```
x = torch.randn(4096, 4096)
y = torch.randn(192, 4096, 1)
z = torch.matmul(x, y)
```

**Before**

```
--------------------------------------------------------------------------------
  n        time(i)         total(B)   useful-heap(B) extra-heap(B)    stacks(B)
--------------------------------------------------------------------------------
 51 18,481,148,839   12,980,531,144   12,978,674,895     1,856,249            0
 52 18,557,183,694   12,980,531,144   12,978,674,895     1,856,249            0
 53 18,633,218,549   12,980,531,144   12,978,674,895     1,856,249            0
 54 18,688,264,583       17,333,504       15,568,202     1,765,302            0
```

**After**

```
--------------------------------------------------------------------------------
  n        time(i)         total(B)   useful-heap(B) extra-heap(B)    stacks(B)
--------------------------------------------------------------------------------
 76  5,124,473,874       96,953,336       95,088,436     1,864,900            0
 77  5,139,417,005       95,591,328       93,726,580     1,864,748            0
 78  5,155,408,421       22,011,848       20,148,194     1,863,654            0
 79  5,171,971,669       21,145,472       19,304,096     1,841,376            0
 80  5,180,817,596       17,866,168       16,086,322     1,779,846            0
 81  5,189,662,538        7,099,760        6,442,537       657,223            0
```

***Timings****

**Test case**

```
x = torch.randn(4096, 4096)
y = torch.randn(192, 4096, 1)

with torch.autograd.profiler.profile(use_cuda=True) as prof:
    z = torch.matmul(x, y)

print(prof)
```

**Before**

```
Self CPU time total: 7.430s
CUDA time total: 7.430s
```

**After**

```
Self CPU time total: 445.886ms
CUDA time total: 445.883ms
```

**Comment by skrah** [comment]
For small matrices the speed is about the same:

```
x = torch.randn(64)
y = torch.randn(2, 64, 1)

with torch.autograd.profiler.profile(use_cuda=True) as prof:
    for i in range(10000):
        z = torch.matmul(x, y)
```

**Before**

```
Self CPU time total: 1.076s
CUDA time total: 1.081s
```

**After**

```
Self CPU time total: 939.033ms
CUDA time total: 940.369ms
```

**Comment by skrah** [comment]
The PR is conservative in not reusing the `out_opt` argument in the recursive call, but I don't see a big difference in the timings (both the old and the new code have the same timings with or without and explicit `out` arg).

**Comment by ezyang** [comment]
Hey @colesbury, do you think you can review this? If not, I'll look.

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 1]
```diff
@@ -428,6 +428,39 @@ Tensor matmul(
...
+    const int64_t p = tensor2.size(-1);
+
+    std::vector<int64_t> perm(dim_tensor2);
+    for (int64_t i = 0; i < dim_tensor2-2; i++) {
+        perm[i] = i;
+    }
+    perm[dim_tensor2-2] = dim_tensor2-1;
+    perm[dim_tensor2-1] = dim_tensor2-2;
+
+    const Tensor t2_T = tensor2.permute(perm);
```
Comment (colesbury) [inline aten/src/ATen/native/LinearAlgebra.cpp]: Can this just be `tensor2.transpose(-1, -2)`?

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 2]
```diff
@@ -428,6 +428,39 @@ Tensor matmul(
...
+    const Tensor t1_T = dim_tensor1 == 2 ? tensor1.t() : tensor1.reshape({n, m}).t();
+    const Tensor res_T = matmul(out_opt, t2_T, t1_T);
+
+    if (dim_tensor1 == 2) {
+      Tensor res = res_T.permute(perm).contiguous();
```
Comment (colesbury) [inline aten/src/ATen/native/LinearAlgebra.cpp]: same here

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 3]
```diff
@@ -428,6 +428,39 @@ Tensor matmul(
...
+      return has_out ? out.set_(res) : res;
+    }
+    else {
+      const IntArrayRef sizes = tensor2.sizes();
```
Comment (colesbury) [inline aten/src/ATen/native/LinearAlgebra.cpp]: This is fine, but you can also write it as:

```
std::vector<int64_t> shape = tensor2.sizes().slice(0, dim_tensor2 - 2).vec();
shape.push_back(p);
```

**Review by colesbury — COMMENTED** [review]
Looks good. Please make sure there are correctness tests in test_torch.py and test_autograd.py that cover this case. Specifically:

dim_tensor1=1, dim_tensor2 >= 3
dim_tensor1=2, dim_tensor2 >= 3

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 1 above)
Comment (skrah) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Yes, that looks nicer.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 3 above)
Comment (skrah) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Thanks, that looks better, too.

**Comment by skrah** [comment]
`test_torch.py` already has tests here (they fail if an error is introduced in the new code):

https://github.com/pytorch/pytorch/blob/8e26759f14c9e4e310e05058fcb687ddc075825c/test/test_torch.py#L4137

I've added some basic `test_autograd.py` tests.


Also I've tested the code with a couple of throwaway scripts like this one, but they may be too long for the unit tests:

```
import torch
import numpy as np

# 2d, 3d
for N in range(1, 20):
  for M in range(1, 20):
    for P in range(1, 20):
      for O in range(1, 20):
        x = torch.arange(N*M).reshape(N, M)
        y = torch.arange(O*M*P).reshape(O, M, P)
        expected = torch.bmm(x.unsqueeze(0).expand(O, N, M), y)
        z = torch.matmul(x, y)
        if not torch.equal(z, expected):
          raise RuntimeError("different results: %s %s %s %s" % (N, M, P, O))
        # Check contiguity flags via numpy.
        ex = np.array(expected, copy=False)
        zz = np.array(z, copy=False)
        if ex.flags != zz.flags or ex[0].flags != zz[0].flags:
          raise RuntimeError("different flags: %s %s %s %s" % (N, M, P, O))

# 1d, 3d
N = 1
for M in range(1, 20):
  for P in range(1, 20):
    for O in range(1, 20):
      x = torch.arange(M)
      y = torch.arange(O*M*P).reshape(O, M, P)
      expected = torch.bmm(x.expand(O, N, M), y).reshape(O, P)
      z = torch.matmul(x, y)
      if not torch.equal(z, expected):
        raise RuntimeError("different results: %s %s %s %s" % (N, M, P, O))
      # Check contiguity flags via numpy.
      ex = np.array(expected, copy=False)
      zz = np.array(z, copy=False)
      if ex.flags != zz.flags or  ex[0].flags != zz[0].flags:
        raise RuntimeError("different flags: %s %s %s %s" % (N, M, P, O))
```

**Comment by skrah** [comment]
@colesbury Thanks for the comments, I think they have all been addressed.

**Comment by ezyang** [comment]
> Also I've tested the code with a couple of throwaway scripts like this one, but they may be too long for the unit tests

That's what the `@slowTest` decorator is for :)

**Comment by skrah** [comment]
> That's what the `@slowTest` decorator is for :)

OK thanks, actually I can add that test with `range(1, 10)`, then it just takes 1s (4s in a debug build).

 Apparently `test_matmul_4d_4d` was banned at some point, it is in a list `THESE_TAKE_WAY_TOO_LONG` and I can no longer find the test itself. :)

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**